search-sentiments-map/routes/trends.js

```diff
@@ -1,35 +1,179 @@
+      trendingSearches = JSON.parse(dailyTrendsJsonString)
+          .default.trendingSearchesDays[0].trendingSearches;
+      trendsByCountry.push(constructCountryTrendsJson(trendingSearches, country.id));
+    }).catch(err => {
+      console.log(err);
+    });
+  }
+
+  saveTrendsAndDeletePrevious(trendsByCountry);
+}
+
+/** Creates a JSON item for trends in the given country. */
+function constructCountryTrendsJson(trendingSearches, countryCode) {
+  let trends = [];
+
+  trendingSearches.forEach(trend => {
+    let articleTitles = [];
+    trend.articles.forEach(article => {
+      articleTitles.push(article.title);
+    })
+    trends.push({
+      topic: trend.title.query,
+      articles: articleTitles,
+      sentimentScore: 0,
+    });
+  })
+
+  return {
+    country: countryCode,
+    trends: trends,
+  }
+}
+
+/**
+ * Delete previous trends and save the current trends in a `trendsEntry` entity
+ * in the Datastore, given the trends JSON organized by country.
+ * Example data structure for a `trendsEntry`:
+ * {timestamp: 111111111,
+    trendsByCountry: [{
+        country: US,
+        trends: [{
+          topic: Donald Trump,
+          articles: [title1, ..., title7],
+          sentimentScore: 0.2,
+          }...
+        ]}, {
+        country: UK,
+        trends: [..., ...]
+      }...
+    ],
+    globalTrends: [{
+        country: US,
+        trendTopic: X,
+    }, ...]
+   }
+ */
+async function saveTrendsAndDeletePrevious(trendsJsonByCountry) {
+  await deleteAncientTrend();
+
+  const trendsEntryKey = datastore.key('TrendsEntry');
+  const trendsEntry = {
+    key: trendsEntryKey,
+    data: {
+      timestamp: Date.now(),
+      trendsByCountry: trendsJsonByCountry,
+      globalTrends: getGlobalTrends(trendsJsonByCountry),
+    },
+  };
+
+  try {
+    await datastore.save(trendsEntry);
+    console.log(`TrendsEntry ${trendsEntryKey.id} created successfully.`);
+  } catch (err) {
+    console.error('ERROR:', err);
+  }
+}
+
+/** 
+ * Delete the oldest trend record if it was saved more than 7 days ago. 
+ */
+async function deleteAncientTrend() {
+  // Query entries in ascending order of the time of creation.
+  const query = datastore.createQuery('TrendsEntry').order('timestamp');
+  const [trendsEntries] = await datastore.runQuery(query);
+
+  if (Date.now() - trendsEntries[0].timestamp > 7 * 24 * 60 * 60000) {
+    const trendsEntryKey = trendsEntries[0][datastore.KEY];
+    await datastore.delete(trendsEntryKey);
+    console.log(`TrendsEntry ${trendsEntryKey.id} deleted.`);
+  }
+}
+
+/** 
+ * Finds the globally trending topics based on trending topics in each country.
+ * Currently returning all topics from the US. TODO(@chenyuz): get a mixture of
+ * topics from different countries.
+ */
+function getGlobalTrends(trendsByCountry) {
+  // Find all trends list(s) whose designated country is the US.
+  let UStrends = trendsByCountry.filter(trends => trends['country'] === 'US');
```

Comment:
Something to think about when it comes time to filter by country, it is faster to filter on the Datastore side so we may want to consider handling getting the top trends on that side.

Reply:
I don't get it. Do you mean to retrieve trends from a specific country from the datastore here? Since the data structure is nested I'm not sure how exactly that can be done.

Reply:
It's not necessary here since it is only done every 12 hours, but yeah we may want to be more efficient when displaying trends for each country if possible.

Reply:
Also why is it faster to filter on the datastore side than filtering the JSON locally?